cmake/bintools/gnu/target_objcopy.cmake

```diff
@@ -0,0 +1,128 @@
+    message(FATAL_ERROR "Both FILE_INPUT and FILE_OUTPUT is required.")
+  endif()
+
+  # Handle stripping
+  set(btOCpyStrip "")
+  if(${BINTOOLS_OBJCOPY_STRIP_ALL})
+    set(btOCpyStrip "-S")
+  elseif(${BINTOOLS_OBJCOPY_STRIP_DEBUG})
+    set(btOCpyStrip "-g")
+  endif()
+
+  # Handle gap filling
+  set(btOCpyGapFill "")
+  if(DEFINED BINTOOLS_OBJCOPY_GAP_FILL)
+    set(btOCpyGapFill "--gap-fill;${BINTOOLS_OBJCOPY_GAP_FILL}")
+  endif()
+
+  # Handle srec len, but only if target output is srec
+  set(btOCpySRecLen "")
+  if(DEFINED BINTOOLS_OBJCOPY_SREC_LEN)
+    if(NOT ${BINTOOLS_OBJCOPY_TARGET_OUTPUT} STREQUAL "srec")
+      message(WARNING "Ignoring srec len, for non srec target: ${BINTOOLS_OBJCOPY_TARGET_OUTPUT}")
+    else()
+      set(btOCpySRecLen "--srec-len;${BINTOOLS_OBJCOPY_SREC_LEN}")
+    endif()
+  endif()
+
+  # Handle Input and Output target types
+  set(btOCpyTargetInp "")
+  if(DEFINED BINTOOLS_OBJCOPY_TARGET_INPUT)
+    set(btOCpyTargetInp "--input-target=${BINTOOLS_OBJCOPY_TARGET_INPUT}")
+  endif()
+  set(btOCpyTargetOut "")
+  if(DEFINED BINTOOLS_OBJCOPY_TARGET_OUTPUT)
+    set(btOCpyTargetOut "--output-target=${BINTOOLS_OBJCOPY_TARGET_OUTPUT}")
+  endif()
+
+  # Handle sections, if any
+  # 1. Section only selection(s)
+  set(btOCpySecOnly "")
+  if(DEFINED BINTOOLS_OBJCOPY_SECTION_ONLY)
+    foreach(secOnly ${BINTOOLS_OBJCOPY_SECTION_ONLY})
+      list(APPEND btOCpySecOnly "--only-section=${secOnly}")
+    endforeach()
+  endif()
+
+  # 2. Section rename selection(s)
+  set(btOCpySecRena "")
+  if(DEFINED BINTOOLS_OBJCOPY_SECTION_RENAME)
+    foreach(secRena ${BINTOOLS_OBJCOPY_SECTION_RENAME})
+      if(NOT ${secRena} MATCHES "^.*=.*$")
+        message(FATAL_ERROR "Malformed section renaming. Must be from=to, have ${secRena}")
+      else()
+        list(APPEND btOCpySecRena "--rename-section;${secRena}")
+      endif()
+    endforeach()
+  endif()
+
+  # 3. Section remove selection(s)
+  set(btOCpySecRem "")
+  if(DEFINED BINTOOLS_OBJCOPY_SECTION_REMOVE)
+    foreach(secRem ${BINTOOLS_OBJCOPY_SECTION_REMOVE})
+      list(APPEND btOCpySecRem "--remove-section=${secRem}")
+    endforeach()
+  endif()
+
+  # Construct the command
+  set(btOCpyCmd COMMAND ${CMAKE_OBJCOPY} ${btOCpyStrip} ${btOCpyGapFill} ${btOCpySRecLen})
+  #   Add input and output target types
+  list(APPEND btOCpyCmd ${btOCpyTargetInp} ${btOCpyTargetOut})
+  #   Add sections
+  list(APPEND btOCpyCmd ${btOCpySecOnly} ${btOCpySecRem} ${btOCpySecRena})
+  #   Add input and output
+  list(APPEND btOCpyCmd ${BINTOOLS_OBJCOPY_FILE_INPUT} ${BINTOOLS_OBJCOPY_FILE_OUTPUT})
+
+  # Place commad on list
+  list(APPEND ${BINTOOLS_OBJCOPY_RESULT_CMD_LIST} ${btOCpyCmd})
+
+  # Add byproduct to list, if list is available
+  if(DEFINED BINTOOLS_OBJCOPY_RESULT_BYPROD_LIST AND DEFINED ${BINTOOLS_OBJCOPY_RESULT_BYPROD_LIST})
+    list(APPEND ${BINTOOLS_OBJCOPY_RESULT_BYPROD_LIST} ${BINTOOLS_OBJCOPY_FILE_OUTPUT})
+  endif()
```

Comment:
I don't see the need for the implementation to append FILE_INPUT to the BYPROD_LIST.

Could it not be up to the user, and not the implementation, to decide whether the FILE_INPUT should be added to byproduct list or not?

I suspect the above three lines will be the same in all implementations.

Reply:
Not entirely sure what you mean here.
It is the FILE_OUTPUT which is put on the BYPROD_LIST, but only if the BYPROD_LIST parameter was actually supplied to the call.
Some calls will provide the by products list, others will not and for some toolchains it may be necessary to add, or not to add, information to the by products list, depending upon how the tool handles the requested operation, perhaps it even needs to put more than one by product on the list, hence the need for having the tool function handle the by product list assignment.

Reply:
Sorry, mixed up FILE_INPUT and FILE_OUTPUT, but my point remains.

> for some toolchains it may be necessary to add, or not to add, information to the by products list

What I am saying is that this is not toolchain-dependent, and therefore should not be in the toolchain implementation.

> perhaps it even needs to put more than one by product on the list, hence the need for having the tool function handle the by product list assignment.

I don't see the need to ever put more items in the BYPROD_LIST than there are items in FILE_OUTPUT, hence why we should not do this in the toolchain implementation.

Reply:
Ok, no problem.
Yes, in essence it is true that the FILE_OUTPUT must always be added to the post_build_byproducts list, and as such it does not need to be added by the binary tools function, but rather by the top CMakeLists.txt. But that said, it should still be possible for the binary tool function to add other by products to the list, even if there is currently no direct use for that.
I will update the PR so the CMakeLists.txt adds the output file to the by products list, but I will keep the list for the binary tool function, so it can add more if needed. 
How does that sound to you?

Reply:
>  it should still be possible for the binary tool function to add other by products to the list, even if there is currently no direct use for that.

We need to be able to imagine some use-case for this, or else we risk having an unnecessary abstraction taking up space, but doing nothing forever.

Reply:
The CMakeLists.txt now adds the output directly to the byproducts list, and the add in the tool functions has been removed.
The use-case I see for why the tool functions needs to be able to add to the byproducts list is, when the implementation outputs some middle file(s), which needs to be kept for reference. So I have still provided the functions with that option.